File: Lecture41/sawalk.cpp
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

inline double std_rand()
{
    return rand() / (RAND_MAX + 1.0);
}

struct Site {               // object to represent lattice site
    int x;                  // x coordinate
    int y;                  // y coordinate
};

int main() {

    cout << " Self-Avoiding Walks on a Square Lattice\n"
         << " ---------------------------------------\n"
         << " Enter number of steps in walk: ";
    int n_steps, n_walks;
    cin >> n_steps;
    cout << " Enter number of walks to generate: ";
    cin >> n_walks;

    int walks = 0;
    int failed_walks = 0;
    double r2av = 0;
    double r4av = 0;

    // generate walks
    while (walks < n_walks) {
        vector<Site> sites;         // set of occupied lattice sites
        Site s;
        s.x = 0;
        s.y = 0;
        sites.push_back(s);
        bool walk_failed = false;

        // loop over desired number of steps
        for (int step = 0; step < n_steps; step++) {

            // take a random step
            double d = std_rand();
            if      (d < 0.25)  ++s.x;   // step East
            else if (d < 0.50)  ++s.y;   // step North
            else if (d < 0.75)  --s.x;   // step West
            else                --s.y;   // step South

            // check whether the site is occupied
            bool occupied = false;
            for (int i = 0; i < sites.size(); i++) {
                if (s.x == sites[i].x && s.y == sites[i].y) {
                    occupied = true;
                    break;
                }
            }
            if (occupied) {
                walk_failed = true;
                break;
            }

            sites.push_back(s);
        }

        if (walk_failed) {
            ++failed_walks;
            continue;
        }

        double r2 = s.x * s.x + s.y * s.y;
        r2av += r2;
        r4av += r2 * r2;
        ++walks;
    }

    r2av /= n_walks;
    r4av /= n_walks;
    double stdDev = sqrt(r4av - r2av * r2av);
    double totalWalks = n_walks + failed_walks;
    double failedPercent = failed_walks / totalWalks * 100.0;
    cout << " Mean square distance <r^2> = " << r2av << "\n"
         << " Standard deviation         = " << stdDev << "\n"
         << " Percentage failed walks    = " << failedPercent
         << endl;
}

File: Lecture41/sawalk.py
import math
import random

print(" Self-Avoiding Walks on a Square Lattice")
print(" ---------------------------------------")
n_steps = int(input(" Enter number of steps in walk: "))
n_walks = int(input(" Enter number of walks to generate: "))

walks = 0
failed_walks = 0
r2_av = 0
r4_av = 0

# generate walks
while walks < n_walks:

    sites = []          # list of occupied lattice sites
    x, y = [0, 0]       # starting site
    sites.append([x, y])
    walk_failed = False

    # loop over desired number of steps
    for step in range(n_steps):

        # take a random step
        d = random.random()
        if d < 0.25:
            x += 1      # step East
        elif d < 0.50:
            y += 1      # step North
        elif d < 0.75:
            x -= 1      # step West
        else:
            y -= 1      # step South

        # check whether the site is occupied
        if [x, y] in sites:
            walk_failed = True
            failed_walks += 1
            break
        else:
            sites.append([x, y])

    if walk_failed:
        continue        # try again

    r2 = x**2 + y**2
    r2_av += r2
    r4_av += r2**2
    walks += 1

r2_av /= n_walks
r4_av /= n_walks
std_dev = math.sqrt(r4_av - r2_av**2)
total_walks = n_walks + failed_walks
failed_percent = failed_walks / total_walks * 100.0
print(" Mean square distance <r^2> =", r2_av)
print(" Standard deviation         =", std_dev)
print(" Number of walks attempted  =", total_walks)
print(" Percentage failed walks    =", failed_percent)
